File: R/test_look_and_say.R
#!/usr/bin/env Rscript

library(rbenchmark)

args = commandArgs(TRUE)

# Get the order n of the look and say sequence from the command line.
if (length(args) == 0) {
   stop("At least one argument must be supplied (dimension).n", call.=FALSE)
} else if (length(args) == 1) {
   n <- as.integer(args[1])
}

LookAndSaySequence <- function(start.sequence.integer, return.an.int=FALSE) {
  # Determine the look and say sequence of a given order, starting with the string x.
  #
  # Args:
  #   start.sequence.integer: The initial sequence of integers used to find the nth look and say sequence.
  #   return.an.int: If TRUE, convert the result to an integer. Otherwise, the result is a string.
  #
  # Returns:
  #   new.string: The completed look and say sequence of order n as a string or an integer, given return.an.int.
  sequence.string <- unlist(strsplit(as.character(start.sequence.integer), ""))  
  rle.sequence.string <- rle(sequence.string)  # Get run length encoding. 

  #Form new string.
  odds <- as.character(rle.sequence.string$lengths)
  evens <- rle.sequence.string$values
  new.string <- as.vector(rbind(odds, evens))

  new.string <- paste(new.string, collapse="")  # Collapse to scalar.
  if (return.an.int) as.integer(new.string) else new.string  # Convert to number, if desired.
}


cat('-----------------------------------------\n')
cat('  Look and say sequence of order: ', n,         '\n')
cat('-----------------------------------------\n')

benchmark("Look_and_Say" = {
            x <- 1223334444
            if (n > 1) {
              for (i in 2:n) {
                x <- LookAndSaySequence(x)
                #print(x)
              }
            }
          },
          replications = 1,
          columns = c("test", "replications", "elapsed", "relative", "user.self", "sys.self"))

File: Python/test_look_and_say.py
#!/usr/bin/env python

from __future__ import print_function

import sys
import benchmark_decorator as dectimer

#--------------------------------
# Function: look_and_say_sequence
#--------------------------------
@dectimer.bench_time(3)
def look_and_say_sequence(starting_sequence, n):
    """
      Construct the look and say sequence of order n and starting 
      sequence starting_sequence (string)
    """
    i = 0

    while i < n:
        if i == 0:
            current_sequence = starting_sequence
        else:
            count = 1
            temp_sequence = ""
            for j in range(1, len(current_sequence)):
                if current_sequence[j] == current_sequence[j-1]:
                    count += 1
                else:
                    temp_sequence = temp_sequence + str(count) \
                                    + current_sequence[j-1]
                    count = 1
            temp_sequence = temp_sequence + str(count)\
                + current_sequence[len(current_sequence) - 1]

            current_sequence = temp_sequence
        i += 1
    return current_sequence


if len(sys.argv) < 2:
    print('Usage:')
    print('     python ' + sys.argv[0] + ' N')
    print('Please specify a number.')
    sys.exit()

N = int(sys.argv[1])

print('Look of say sequence of order: ', N)

seq = look_and_say_sequence("1223334444", N)

#print('Sequence of order ', N, ': ', seq)
print(' ')
